Repeated querySelector('[attrib=values]') calls are slow.

mattwoodrow opened the pull request:
#### 74ea666b2e01f4766dd835ab2716cd60d4eed575
<pre>
Repeated querySelector(&apos;[attrib=values]&apos;) calls are slow.
<a href="https://bugs.webkit.org/show_bug.cgi?id=290590">https://bugs.webkit.org/show_bug.cgi?id=290590</a>
&lt;<a href="https://rdar.apple.com/148064814">rdar://148064814</a>&gt;

Reviewed by Matthieu Dubet.

Each call has to do repeat a tree search from the start, even though the results
are frequently close together.

When we do one of these searches, cache the first element with a potentially
matching attribute.  For repeated calls (without any relevant tree mutations),
reduce the search space by starting at the cached element.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::cachedFirstContainerWithAttribute const):
(WebCore::Document::setCachedFirstContainerWithAttribute):
(WebCore::Document::attributeAddedToElement):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::notifyAttributeChanged):
(WebCore::Element::insertedIntoAncestor):
* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::SelectorDataList::executeSingleAttributeExactSelectorData const):

Canonical link: <a href="https://commits.webkit.org/293075@main">https://commits.webkit.org/293075@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db2d74c8995243f37025b4ca207e290aeb4e9a2f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97842 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17467 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7686 "Built successfully") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102954 "Failed to checkout and rebase branch from PR 43156") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48371 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99887 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17761 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25920 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/102954 "Failed to checkout and rebase branch from PR 43156") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31681 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100845 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13446 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/88426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/102954 "Failed to checkout and rebase branch from PR 43156") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13223 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6351 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47814 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6424 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/105334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24922 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/105334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25294 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84599 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/105334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20921 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27564 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/5252 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18538 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24883 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/30052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24705 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28019 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26279 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Andrew-Fryer commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2977,6 +2980,11 @@ Node::InsertedIntoAncestorResult Element::insertedIntoAncestor(InsertionType ins
+            for (auto& attribute : attributes())
+                document().attributeAddedToElement(attribute.name());
+        }
+
```
It looks like this patch currently covers the following cases:
- attribute added to element
- element inserted into dom

What about these cases?
- attribute removed from element
- element removed from dom

mattwoodrow replied:
Removing an attribute doesn't need to do anything, since the existing search start point is still always valid, it just might not be optimal (but still better than invalidating the cache). What I can do, is make future searches update the cached Element, so that this fixes itself if it gets used.

Removing an element is a good point. I thought WeakPtr would be sufficient, but we can't start the tree walk at a position that isn't in the tree. Fixed.

Andrew-Fryer commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2224,6 +2224,9 @@ void Element::notifyAttributeChanged(const QualifiedName& name, const AtomString
     }
+
+    if (isConnected() && oldValue == nullAtom())
+        document().attributeAddedToElement(name);
```
I wonder if we'd get a benefit from checking if the parent of this element is the same element that's already in the cache (in which case I think we don't need to invalidate the cache).
I'd image that siblings might fight here ... perhaps we should pass `this` into `attributeAddedToElement` and do the check in there?
I think the check should just be a pointer compare of data inlined in `this`, so I think overhead would be low.

Andrew-Fryer commented on Source/WebCore/dom/SelectorQuery.cpp:
```diff
@@ -437,8 +437,16 @@ ALWAYS_INLINE void SelectorDataList::executeSingleAttributeExactSelectorData(con
+
+    auto elementDescendants = descendantsOfType<Element>(const_cast<ContainerNode&>(rootNode));
+    for (auto it = cachedContainer ? elementDescendants.beginAt(*cachedContainer) : elementDescendants.begin(); it; ++it) {
+        Ref element = *it;
```
May I ask why we store the container rather than the element?
It makes this code elegant, but I think storing starting iteration at the element rather than its parent would be a bit faster.
I'm guessing the reason is that `descendantsOfType(Element)` doesn't exist... (I don't see it in `Source/WebCore/dom`)

mattwoodrow replied:
It was vestigial from an earlier (and invalid) revision where I needed to look at earlier siblings than the matched one.

You're right, it's both simpler and more precise to just store the matched element, have done that.

Andrew-Fryer replied:
(FYI: the commit msg is stale now)

Andrew-Fryer commented on Source/WebCore/dom/Document.h:
```diff
@@ -2346,6 +2353,8 @@ class Document
     RefPtr<ViewTransition> m_activeViewTransition;
 
+    std::optional<std::pair<QualifiedName, WeakPtr<Element, WeakPtrImplWithEventTargetData>>> m_cachedFirstElementWithAttribute;
+
```
I wonder if storing elements for N LRU attributes would give more benefit for some small N.

mattwoodrow replied:
It wouldn't be difficult to switch to using `wtf::TinyLRUCache` at least.

It does make `Document` bigger though, so I think we should wait for data showing that it's a common pattern to do repeated searches with multiple attributes (and what value of N would be needed to improve them).

mdubet commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -2224,6 +2224,9 @@ void Element::notifyAttributeChanged(const QualifiedName& name, const AtomString
             cache->deferAttributeChangeIfNeeded(*this, name, oldValue, newValue);
     }
+
+    if (isConnected() && oldValue == nullAtom())
```
Should we move that inside the `if (oldValue != newValue)` like the other `invalidate...Caches...ForAttribute()` call ?

mdubet commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -11430,6 +11430,30 @@ void Document::securityOriginDidChange()
+        m_cachedFirstElementWithAttribute = std::nullopt;
+}
+
+void Document::elementDisconnectedFromDocument(Element& element)
```
`const Element&`

mdubet reviewed (approved):
💯

mattwoodrow committed:
Repeated querySelector('[attrib=values]') calls are slow.
https://bugs.webkit.org/show_bug.cgi?id=290590
<rdar://148064814>

Reviewed by Matthieu Dubet.

Each call has to do repeat a tree search from the start, even though the results
are frequently close together.

When we do one of these searches, cache the first element with a potentially
matching attribute.  For repeated calls (without any relevant tree mutations),
reduce the search space by starting at the cached element.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::cachedFirstContainerWithAttribute const):
(WebCore::Document::setCachedFirstContainerWithAttribute):
(WebCore::Document::attributeAddedToElement):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::notifyAttributeChanged):
(WebCore::Element::insertedIntoAncestor):
* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::SelectorDataList::executeSingleAttributeExactSelectorData const):

Canonical link: https://commits.webkit.org/293075@main
